ratis-server-api/src/main/java/org/apache/ratis/statemachine/StateMachine.java

```diff
@@ -189,6 +189,12 @@ default void notifyGroupRemove() {}
      * @param failedEntry The failed log entry, if there is any.
      */
     default void notifyLogFailed(Throwable cause, LogEntryProto failedEntry) {}
+
+    /**
+     * Notify the {@link StateMachine} that the progress of install snapshot is
+     * completely done. Could trigger the cleanup of snapshots.
+     */
+    default void notifyInstallSnapshotFinished(RaftProtos.InstallSnapshotResult result) {}
```

Comment:
Since this method is for the state machine to clean up, it is useful to pass also the snapshot index.

BTW, let's rename this to notifySnapshotInstalled as below.
```
    default void notifySnapshotInstalled(InstallSnapshotResult result, long snapshotIndex) {}
```

Reply:
> @Xushaohong , Since the new method is for the state machine to clean up, it is useful to pass also the snapshot index; see https://issues.apache.org/jira/secure/attachment/13044817/647_review.patch (apply on top of the change)

Only one place modified, where the initial value should be consistent as -1
`private final AtomicLong installedIndex = new AtomicLong(INVALID_LOG_INDEX);`